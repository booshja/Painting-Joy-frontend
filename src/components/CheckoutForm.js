// dependencies
import React, { useContext, useRef } from "react";
import { useHistory } from "react-router";
import { useForm } from "react-hook-form";
import styled from "styled-components";
import axios from "axios";
import ReCAPTCHA from "react-google-recaptcha";
// context
import CartContext from "../context/CartContext";
// assets
import stateCodes from "../assets/stateCodes.js";

const StyledForm = styled.form`
    display: flex;
    flex-direction: column;
    align-items: flex-start;
`;

const StyledLabel = styled.label`
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    margin-bottom: 0.25rem;
`;

const StyledInput = styled.input`
    border: 2px solid #207070;
    border-radius: 4px;
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    margin-bottom: 1rem;

    &::-webkit-outer-spin-button,
    &::-webkit-inner-spin-button {
        -webkit-appearance: none;
        margin: 0;
    }
    &[type="number"] {
        -moz-appearance: textfield;
    }
`;

const StyledSelect = styled.select`
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    border: 2px solid #207070;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 1rem;
`;

const StyledBtnContainer = styled.div`
    display: flex;
    align-items: center;
    align-self: flex-end;
    margin-top: 1rem;
`;

const StyledContBtn = styled.button`
    background-color: #207070;
    border: none;
    border-radius: 4px;
    padding: 0.5px 8px;
    color: #ffffff;
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.4rem;
    letter-spacing: 1px;
    cursor: pointer;
    border: 2px solid #207070;
    margin-left: 1rem;

    &:hover {
        background-color: #ffffff;
        color: #207070;
    }
`;

const StyledCancelBtn = styled.button`
    background-color: #ffffff;
    border: 2px solid #990000;
    border-radius: 4px;
    padding: 0.5px 8px;
    color: #990000;
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.4rem;
    letter-spacing: 1px;
    cursor: pointer;
    border: 2px solid #990000;

    &:hover {
        background-color: #990000;
        color: #ffffff;
    }
`;

const StyledError = styled.p`
    color: #990000;
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.2rem;
    letter-spacing: 1px;
    padding: 0.5rem;
`;

const CheckoutForm = ({ nextStep, amount }) => {
    // set up react-hook-form
    const {
        register,
        handleSubmit,
        formState: { errors },
    } = useForm();
    // set up history
    const history = useHistory();
    // set up context
    const { orderId } = useContext(CartContext);
    // set up ref
    const reRef = useRef();

    const handleDataSubmit = async (data) => {
        // get recaptcha token
        const token = await reRef.current.executeAsync();
        reRef.current.reset();

        // when form submits, send data to back end, trigger next step
        try {
            await axios.post(
                process.env.REACT_APP_BACKEND_URL +
                    `orders/order/${orderId}/info`,
                {
                    name: data.fullName,
                    email: data.email,
                    street: data.street,
                    unit: data.unit,
                    city: data.city,
                    stateCode: data.stateCode,
                    zipcode: +data.zipcode,
                    phone: +data.phone,
                    amount: +amount,
                    token,
                }
            );
            nextStep(data.email);
        } catch (err) {
            console.log("Checkout Error:", err);
            // history.push("/store/order/cancel");
        }
    };

    const handleCancel = async (e) => {
        // on cancel, abort order and send user to order cancelled page
        e.preventDefault();
        try {
            await axios.delete(
                process.env.REACT_APP_BACKEND_URL +
                    `orders/order/${orderId}/abort`
            );
            history.push("/store/order/cancel");
        } catch (err) {
            console.log("Checkout Abort Error:", err);
            history.push("/store/order/cancel");
        }
    };

    return (
        <StyledForm
            onSubmit={handleSubmit((data) => {
                handleDataSubmit(data);
            })}
        >
            <ReCAPTCHA
                sitekey={process.env.REACT_APP_RECAPTCHA_SITE_KEY}
                size="invisible"
                ref={reRef}
            />
            <StyledLabel htmlFor="fullName">Name:</StyledLabel>
            <StyledInput
                type="text"
                placeholder="Full Name"
                id="fullName"
                required
                {...register("fullName", { required: "Name is required" })}
            />
            {errors.name && (
                <StyledError>{errors.fullName.message}</StyledError>
            )}
            <StyledLabel htmlFor="email">Email Address:</StyledLabel>
            <StyledInput
                type="text"
                placeholder="Email Address"
                id="email"
                required
                {...register("email", {
                    required: "Email address is required",
                    type: "email",
                })}
            />
            {errors.email && <StyledError>{errors.email.message}</StyledError>}
            <StyledLabel htmlFor="street">Street:</StyledLabel>
            <StyledInput
                type="text"
                placeholder="Street"
                id="street"
                required
                {...register("street", {
                    required: "Street address is required",
                })}
            />
            {errors.unit && <StyledError>{errors.unit.message}</StyledError>}
            <StyledLabel htmlFor="street">Unit:</StyledLabel>
            <StyledInput
                type="text"
                placeholder="Unit (Optional)"
                id="unit"
                {...register("unit")}
            />
            <StyledLabel htmlFor="city">City:</StyledLabel>
            <StyledInput
                type="text"
                placeholder="City"
                id="city"
                required
                {...register("city", {
                    required: "City is required",
                })}
            />
            {errors.city && <StyledError>{errors.city.message}</StyledError>}
            <StyledLabel htmlFor="stateCode">State:</StyledLabel>
            <StyledSelect {...register("stateCode")}>
                {stateCodes.map(({ code, state }) => (
                    <option value={code} key={code}>
                        {state}
                    </option>
                ))}
                {errors.stateCode && (
                    <StyledError>{errors.stateCode.message}</StyledError>
                )}
            </StyledSelect>
            <StyledLabel htmlFor="zipcode">Zipcode: (ex: 98101)</StyledLabel>
            <StyledInput
                type="number"
                placeholder="Zipcode"
                id="zipcode"
                max="99999"
                min="10000"
                pattern={/ \d{5} /}
                required
                {...register("zipcode", {
                    required: "Zipcode is required",
                    max: "Zipcode must be 5 digits",
                    min: "Zipcode must be 5 digits",
                    pattern: "Zipcode must be 5 digits",
                })}
            />
            {errors.zipcode && (
                <StyledError>{errors.zipcode.message}</StyledError>
            )}
            <StyledLabel htmlFor="phone">
                Phone Number: (ex: 5555555555)
            </StyledLabel>
            <StyledInput
                type="number"
                placeholder="Phone Number"
                id="phone"
                min="1000000000"
                max="9999999999"
                pattern={/ \d{10} /}
                required
                {...register("phone", {
                    required: "Phone number is required",
                    pattern: "Phone number must be 10 digits",
                    min: "Phone number must be 10 digits",
                    max: "Phone number must be 10 digits",
                })}
            />
            {errors.phone && <StyledError>{errors.phone.message}</StyledError>}
            <StyledBtnContainer>
                <StyledCancelBtn onClick={(e) => handleCancel(e)}>
                    Cancel
                </StyledCancelBtn>
                <StyledContBtn type="submit">Continue</StyledContBtn>
            </StyledBtnContainer>
        </StyledForm>
    );
};

export default CheckoutForm;
