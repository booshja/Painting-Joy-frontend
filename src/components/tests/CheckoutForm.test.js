import { render } from "@testing-library/react";
import CheckoutForm from "../CheckoutForm";
import CartContext from "../../context/CartContext";

/** Smoke Test */
it("renders without crashing", () => {
    render(
        <CartContext.Provider value={{ orderId: "orderId123" }}>
            <CheckoutForm nextStep={null} amount="129.99" />
        </CartContext.Provider>
    );
});

/** Snapshot Test */
it("matches snapshot", () => {
    const { asFragment } = render(
        <CartContext.Provider value={{ orderId: "orderId123" }}>
            <CheckoutForm nextStep={null} amount="129.99" />
        </CartContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});

/** RTL Tests */
it("renders correctly", () => {
    const { getByLabelText, getAllByRole } = render(
        <CartContext.Provider value={{ orderId: "orderId123" }}>
            <CheckoutForm nextStep={null} amount="129.99" />
        </CartContext.Provider>
    );

    getByLabelText("Name:");
    getByLabelText("Email Address:");
    getByLabelText("Street:");
    getByLabelText("Unit:");
    getByLabelText("City:");
    getByLabelText("State:");
    getByLabelText("Zipcode: (ex: 98101)");
    getByLabelText("Phone Number: (ex: 5555555555)");
    const buttons = getAllByRole("button");

    expect(buttons[0]).toHaveTextContent("Cancel");
    expect(buttons[1]).toHaveTextContent("Continue");
});
