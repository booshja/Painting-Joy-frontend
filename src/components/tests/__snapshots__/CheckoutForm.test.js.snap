// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-bdnxRM lgoygW"
  >
    <div />
    <label
      class="sc-gtsrHT iZAmyw"
      for="fullName"
    >
      Name:
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="fullName"
      name="fullName"
      placeholder="Full Name"
      required=""
      type="text"
    />
    <label
      class="sc-gtsrHT iZAmyw"
      for="email"
    >
      Email Address:
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="email"
      name="email"
      placeholder="Email Address"
      required=""
      type="text"
    />
    <label
      class="sc-gtsrHT iZAmyw"
      for="street"
    >
      Street:
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="street"
      name="street"
      placeholder="Street"
      required=""
      type="text"
    />
    <label
      class="sc-gtsrHT iZAmyw"
      for="street"
    >
      Unit:
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="unit"
      name="unit"
      placeholder="Unit (Optional)"
      type="text"
    />
    <label
      class="sc-gtsrHT iZAmyw"
      for="city"
    >
      City:
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="city"
      name="city"
      placeholder="City"
      required=""
      type="text"
    />
    <label
      class="sc-gtsrHT iZAmyw"
      for="stateCode"
    >
      State:
    </label>
    <select
      class="sc-hKFxyN dDjeJq"
      id="stateCode"
      name="stateCode"
    >
      <option
        value="AL"
      >
        Alabama
      </option>
      <option
        value="AK"
      >
        Alaska
      </option>
      <option
        value="AZ"
      >
        Arizona
      </option>
      <option
        value="AR"
      >
        Arkansas
      </option>
      <option
        value="CA"
      >
        California
      </option>
      <option
        value="CO"
      >
        Colorado
      </option>
      <option
        value="CT"
      >
        Connecticut
      </option>
      <option
        value="DE"
      >
        Delaware
      </option>
      <option
        value="DC"
      >
        District of Columbia
      </option>
      <option
        value="FL"
      >
        Florida
      </option>
      <option
        value="GA"
      >
        Georgia
      </option>
      <option
        value="HI"
      >
        Hawaii
      </option>
      <option
        value="ID"
      >
        Idaho
      </option>
      <option
        value="IL"
      >
        Illinois
      </option>
      <option
        value="IN"
      >
        Indiana
      </option>
      <option
        value="IA"
      >
        Iowa
      </option>
      <option
        value="KS"
      >
        Kansas
      </option>
      <option
        value="KY"
      >
        Kentucky
      </option>
      <option
        value="LA"
      >
        Louisiana
      </option>
      <option
        value="ME"
      >
        Maine
      </option>
      <option
        value="MD"
      >
        Maryland
      </option>
      <option
        value="MA"
      >
        Massachusetts
      </option>
      <option
        value="MI"
      >
        Michigan
      </option>
      <option
        value="MN"
      >
        Minnesota
      </option>
      <option
        value="MS"
      >
        Mississippi
      </option>
      <option
        value="MO"
      >
        Missouri
      </option>
      <option
        value="MT"
      >
        Montana
      </option>
      <option
        value="NE"
      >
        Nebraska
      </option>
      <option
        value="NV"
      >
        Nevada
      </option>
      <option
        value="NH"
      >
        New Hampshire
      </option>
      <option
        value="NJ"
      >
        New Jersey
      </option>
      <option
        value="NM"
      >
        New Mexico
      </option>
      <option
        value="NY"
      >
        New York
      </option>
      <option
        value="NC"
      >
        North Carolina
      </option>
      <option
        value="ND"
      >
        North Dakota
      </option>
      <option
        value="OH"
      >
        Ohio
      </option>
      <option
        value="OK"
      >
        Oklahoma
      </option>
      <option
        value="OR"
      >
        Oregon
      </option>
      <option
        value="PA"
      >
        Pennsylvania
      </option>
      <option
        value="RI"
      >
        Rhode Island
      </option>
      <option
        value="SC"
      >
        South Carolina
      </option>
      <option
        value="SD"
      >
        South Dakota
      </option>
      <option
        value="TN"
      >
        Tennessee
      </option>
      <option
        value="TX"
      >
        Texas
      </option>
      <option
        value="UT"
      >
        Utah
      </option>
      <option
        value="VT"
      >
        Vermont
      </option>
      <option
        value="VA"
      >
        Virginia
      </option>
      <option
        value="WA"
      >
        Washington
      </option>
      <option
        value="WV"
      >
        West Virginia
      </option>
      <option
        value="WI"
      >
        Wisconsin
      </option>
      <option
        value="WY"
      >
        Wyoming
      </option>
    </select>
    <label
      class="sc-gtsrHT iZAmyw"
      for="zipcode"
    >
      Zipcode: (ex: 98101)
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="zipcode"
      max="99999"
      min="10000"
      name="zipcode"
      pattern="/ \\\\d{5} /"
      placeholder="Zipcode"
      required=""
      type="number"
    />
    <label
      class="sc-gtsrHT iZAmyw"
      for="phone"
    >
      Phone Number: (ex: 5555555555)
    </label>
    <input
      class="sc-dlnjwi cVNoyy"
      id="phone"
      max="9999999999"
      min="1000000000"
      name="phone"
      pattern="/ \\\\d{10} /"
      placeholder="Phone Number"
      required=""
      type="number"
    />
    <div
      class="sc-eCApnc cOzwgC"
    >
      <button
        class="sc-gKAaRy ikZoHU"
      >
        Cancel
      </button>
      <button
        class="sc-jSFjdj iAbnik"
        type="submit"
      >
        Continue
      </button>
    </div>
  </form>
</DocumentFragment>
`;
